types/src/crypto/asymmetric_key.rs

```diff
@@ -133,15 +133,19 @@ impl SecretKey {
     pub const SECP256K1_LENGTH: usize = SECP256K1_SECRET_KEY_LENGTH;
 
     /// Constructs a new Ed25519 variant from a byte array.
-    pub fn ed25519(bytes: [u8; Self::ED25519_LENGTH]) -> Self {
-        // safe to unwrap as `SecretKey::from_bytes` can only fail if the provided slice is the
-        // wrong length.
-        SecretKey::Ed25519(ed25519_dalek::SecretKey::from_bytes(&bytes).unwrap())
+    pub fn ed25519(bytes: [u8; Self::ED25519_LENGTH]) -> Result<Self, Error> {
+        // while it is safe to unwrap as `SecretKey::from_bytes` can only fail if the provided slice
+        // is the wrong length.
+        // We should be disciplined about unwrapping in constructors.
+        match ed25519_dalek::SecretKey::from_bytes(&bytes) {
+            Ok(key) => Ok(SecretKey::Ed25519(key)),
+            Err(error) => Err(Error::AsymmetricKey(error.to_string())),
+        }
     }
 
     /// Constructs a new secp256k1 variant from a byte array.
-    pub fn secp256k1(bytes: [u8; Self::SECP256K1_LENGTH]) -> Self {
-        SecretKey::Secp256k1(bytes)
+    pub fn secp256k1(bytes: [u8; Self::SECP256K1_LENGTH]) -> Result<Self, Error> {
+        Ok(SecretKey::Secp256k1(bytes))
```

Comment:
This has to return an error if bytes is all zeros (that's an invalid private key for Secp256k1)

Reply:
I think if we call the k256 private key constructor here it'll throw that error for us (we'll need to have a `From` impl for the `Error` it thows however).

Reply:
I agree about using the `k256::SecretKey::from_bytes()` constructor to validate these bytes, but for the same reason as my comment on the Ed25519 constructor, I'd avoid providing a `From` impl for that error type.